tools/hardware/utils/memory.py

```diff
@@ -26,6 +26,38 @@ def visitor(node: WrappedNode):
     return regions
 
 
+def merge_memory_regions(regions: Set[Region]) -> Set[Region]:
+    ''' Check all region and merge adjacent ones '''
+    all_regions = []
+    idx = 0
+    for region in regions:
+        all_regions.append(dict(idx=idx, region=region, right_adj=None, left_adj=None))
+        idx += 1
+
+    # Find all right contiguous regions
+    for dreg in all_regions:
+        for dnreg in all_regions[dreg['idx']+1:]:
```

Comment:
This assume that all_regions is sorted?

Reply:
It seems important to keep separated the device memory regions and untyped memory regions. All memory regions are inspected, but contiguity is considered in respect to region owner. And as a result, a memory region spanning over 2 buses would not get merged.